Create MCP documentation (#213)
DefangLabs/defang-docs: docs/concepts/mcp.md

---
title: MCP
description: Model Context Protocol (MCP) server with built-in tools to allow users to deploy with Defang through a supported IDE.
sidebar-position: 200
---

# Model Context Protocol (MCP)

> Defang has a [Model Context Protocol (MCP) server](https://github.com/DefangLabs/defang) with built-in tools to allow users to deploy and manage cloud services through a

Author: Needs link to the mcp folder in /defang repo

---
title: MCP
description: Model Context Protocol (MCP) server with built-in tools to allow users to deploy with Defang through a supported IDE.
sidebar-position: 200
---

# Model Context Protocol (MCP)

> Defang has a [Model Context Protocol (MCP) server](https://github.com/DefangLabs/defang) with built-in tools to allow users to deploy and manage cloud services through a

Reviewer: ```suggestion
The Defang [Model Context Protocol (MCP) server](https://github.com/DefangLabs/defang) with built-in tools to allow users to deploy and manage cloud services through a
```

supported IDE.

Using this MCP server with an IDE will enable the AI coding agent (e.g. Copilot) to use Defang tools and resources to perform tasks, such as deploying a service to the cloud. This means you can now use Defang with IDE-integrated AI coding agents.

![Defang MCP Server Diagram](/img/mcp-concept/diagram.png)

> For more detail on MCP architecture, visit the [official MCP documentation](https://modelcontextprotocol.io/introduction).

Reviewer: ```suggestion
For more details about MCP architecture, visit the [official MCP documentation](https://modelcontextprotocol.io/introduction).
```

:::info
The MCP Server currently supports deployments to [Defang Playground](/docs/providers/playground). We plan to support Defang BYOC in future updates.
:::

## Installation

> First, make sure you have the [npm package manager](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm) installed. Connect the MCP server with your IDE by running the following command in your terminal:

Reviewer 2: Does this install npx also?

Author: yes

```bash
npx -y defang mcp setup --client=<your-ide>
```

Replace `<your-ide>` with the name of your preferred IDE. See our list of [Supported IDEs](#supported-ides). After setup, you can start the MCP server with the command:

```bash

> npx -y defang mcp serve

Reviewer 2: serve or server?

Author: serve

```

Once the server is running, you can access the Defang MCP tools directly through the AI agent chat in your IDE.

## Supported IDEs

### Cursor

```bash
npx -y defang mcp setup --client=cursor
```

### Windsurf

```bash
npx -y defang mcp setup --client=windsurf
```

### VSCode

```bash
npx -y defang mcp setup --client=vscode
```

### Claude Desktop

(While this is not an IDE in the traditional sense, it can support MCP servers.)

```bash
npx -y defang mcp setup --client=claude
```

## MCP Tools

Below are the tools available in the Defang MCP Server.

### `deploy`

> The `deploy` tool scans your project directory for Dockerfiles and `compose.yaml` files, then deploys the detected service(s) using Defang. You can monitor the deployment process in the Defang Portal.

Reviewer 2: should we mention here it deploys to the Defang Playground?

Author: okay, i'll move the following sentence to that section instead. 

"The MCP Server currently supports deployments to [Defang Playground](/docs/providers/playground). We plan to support Defang BYOC in future updates."

Author: Sure, I can repeat the info notice in that section too

---
title: MCP
description: Model Context Protocol (MCP) server with built-in tools to allow users to deploy with Defang through a supported IDE.
sidebar-position: 200
---

# Model Context Protocol (MCP)

> 
> The Defang [Model Context Protocol (MCP) server](https://github.com/DefangLabs/defang) includes built-in tools to allow users to deploy and manage cloud services through a
> supported IDE.

Reviewer: ```suggestion

The Defang [Model Context Protocol (MCP) server](https://github.com/DefangLabs/defang) includes built-in tools to allow users to deploy and manage cloud services through a supported IDE.
```

```

Replace `<your-ide>` with the name of your preferred IDE. See our list of [Supported IDEs](#supported-ides). After setup, you can start the MCP server with the command:

```bash
npx -y defang mcp serve
```

Once the server is running, you can access the Defang MCP tools directly through the AI agent chat in your IDE.

## Supported IDEs

### Cursor

```bash
npx -y defang mcp setup --client=cursor
```

### Windsurf

```bash
npx -y defang mcp setup --client=windsurf
```

### VSCode

```bash
npx -y defang mcp setup --client=vscode
```

### Claude Desktop

(While this is not an IDE in the traditional sense, it can support MCP servers.)

```bash
npx -y defang mcp setup --client=claude
```

## MCP Tools

Below are the tools available in the Defang MCP Server.

### `deploy`

The `deploy` tool scans your project directory for Dockerfiles and `compose.yaml` files, then deploys the detected service(s) using Defang. You can monitor the deployment process in the Defang Portal.

:::info
The MCP Server currently supports deployments to [Defang Playground](/docs/providers/playground). We plan to support Defang BYOC in future updates.
:::

### `services`

> The `services` tool displays the details of all your services that are currently deployed with Defang. It shows the Service Name, Deployment ID, Public URL and Service Status. If there are no services found, it will display an appropriate message.

Reviewer: ```suggestion
The `services` tool displays the details of all your services that are currently deployed in your application with Defang. It shows the Service Name, Deployment ID, Public URL and Service Status. If there are no services found, it will display an appropriate message.
```

### `destroy`

Given a project name or directory, the `destroy` tool identifies any services deployed with Defang and terminates them. If no services are found, it will display an appropriate message.

:::info
This page is a guide to the Defang MCP Server detailing its installation, tools, and usage. If you are looking for an example of how you can deploy a MCP project with Defang, please instead refer to our [MCP sample application](https://github.com/DefangLabs/samples/tree/main/samples/mcp).
:::

## Example Prompts

> After connecting the Defang MCP Server to your IDE using an installation method, you can type in prompts in your chat to invoke the AI agent to use an MCP tool(s).

Reviewer: ```suggestion
After connecting the Defang MCP Server to your IDE using an installation method, you can type in prompts in your chat to invoke the AI agent to use any MCP tool(s).
```

---
title: MCP
description: Model Context Protocol (MCP) server with built-in tools to allow users to deploy with Defang through a supported IDE.
sidebar-position: 200
---

# Model Context Protocol (MCP)

> The Defang [Model Context Protocol (MCP) server](https://github.com/DefangLabs/defang) includes built-in tools to allow users to deploy and manage cloud services through a supported IDE.

Author: ```suggestion
The Defang [Model Context Protocol (MCP) server](https://github.com/DefangLabs/defang/tree/main/src/pkg/mcp) includes built-in tools to allow users to deploy and manage cloud services through a supported IDE.
```
